src/find/matchers/logical_matchers.rs

```diff
@@ -0,0 +1,477 @@
+        AndMatcher { submatchers: submatchers }
+    }
+}
+
+
+impl Matcher for AndMatcher {
+    /// Returns true if all sub-matchers return true. Short-circuiting does take
+    /// place. If the nth sub-matcher returns false, then we immediately return
+    /// and don't make any further calls.
+    fn matches(&self, dir_entry: &DirEntry, matcher_io: &mut MatcherIO) -> bool {
+        self.submatchers.iter().all(|ref x| x.matches(dir_entry, matcher_io))
+    }
+
+    fn has_side_effects(&self) -> bool {
+        self.submatchers.iter().any(|ref x| x.has_side_effects())
+    }
+}
+
+pub struct AndMatcherBuilder {
+    submatchers: Vec<Box<Matcher>>,
+}
+
+impl AndMatcherBuilder {
+    pub fn new() -> AndMatcherBuilder {
+        AndMatcherBuilder { submatchers: Vec::new() }
+    }
+
+    pub fn new_and_condition(&mut self, matcher: Box<Matcher>) {
+        self.submatchers.push(matcher);
+    }
+
+    /// Builds a Matcher: consuming the builder in the process.
+    pub fn build(mut self) -> Box<Matcher> {
+        // special case. If there's only one submatcher, just return that directly
+        if self.submatchers.len() == 1 {
+            // safe to unwrap: we've just checked the size
+            return self.submatchers.pop().unwrap();
+        }
+        let matcher = Box::new(AndMatcher::new(self.submatchers));
+        self.submatchers = Vec::new();
+        matcher
+    }
+}
+
+
+
+/// This matcher contains a collection of other matchers. A file matches
+/// if it matches any of the contained sub-matchers. For sub-matchers that have
+/// side effects, the side effects occur in the same order as the sub-matchers
+/// were pushed into the collection.
+pub struct OrMatcher {
+    submatchers: Vec<Box<Matcher>>,
+}
+
+impl OrMatcher {
+    pub fn new(submatchers: Vec<Box<Matcher>>) -> OrMatcher {
+        OrMatcher { submatchers: submatchers }
+    }
+}
+
+
+impl Matcher for OrMatcher {
+    /// Returns true if any sub-matcher returns true. Short-circuiting does take
+    /// place. If the nth sub-matcher returns true, then we immediately return
+    /// and don't make any further calls.
+    fn matches(&self, dir_entry: &DirEntry, matcher_io: &mut MatcherIO) -> bool {
+        self.submatchers.iter().any(|ref x| x.matches(dir_entry, matcher_io))
+    }
+
+    fn has_side_effects(&self) -> bool {
+        self.submatchers.iter().any(|ref x| x.has_side_effects())
+    }
+}
+
+pub struct OrMatcherBuilder {
+    submatchers: Vec<AndMatcherBuilder>,
+}
+
+impl OrMatcherBuilder {
+    pub fn new_and_condition(&mut self, matcher: Box<Matcher>) {
+        // safe to unwrap. submatchers always has at least one member
+        self.submatchers.last_mut().unwrap().new_and_condition(matcher);
+    }
+
+    pub fn new_or_condition(&mut self, arg: &str) -> Result<(), Box<Error>> {
+        if self.submatchers.last().unwrap().submatchers.is_empty() {
+            return Err(From::from(format!("invalid expression; you have used a binary operator \
```

Comment:
I would suggest to define our own `Error` enum so that this can be an enumerant. It's more structured that way. :)